src/lib.rs

```diff
@@ -1,3 +1,162 @@
-#[test]
-fn it_works() {
+#[cfg(test)]
+mod test {
+    use tokens::{ Lexer, LexerOptions };
+
+    #[test]
+    fn lexer_usage() {
+        // build the lexer once for project environment with extensions.
+        let lexer = Lexer::new(LexerOptions::default(), vec![]);
+
+        // use many times.
+        {
+            let source: String = "{{ var }}".into();
+            for token in lexer.tokens(&source) {
```

Comment:
I see the motivation behind the iterator pattern - however it has one drawback: It looses valuable information about 'current state'. The lexer definitely has different states. However you turn it, in a for-loop there will always be one entry-point `Iterator<>::next()` - each call to next() would need to lookup the last state and conditionally jump to the corresponding code (either vtable call or some match-pattern).

My current design does not use the iterator pattern for this reason. I don't need to carry a state at runtime, because it is all encoded statically. Each jump is without dynamic indirection or state-lookup-and-branching. And I might happen to generate more than one token between jumps. For an iterator design we would have to make frequent stops and dynamically store our state instead of static jumps.

This is a trade-off:
* fast static jumps means high troughput in terms of "lexed templates per second" for multiple templates - but you have to wait until a template is parsed as a whole
* iteration means you can build a lexer-parser-pipeline which can better parallelize lexing and parsing for a single template - but will use more cpu-cycles per template.

What will be important in the end? Ideally we will cache compiled templates and only regenerate if something changed (different template, different extensions, etc.).

Therefore I think it is not super important - but my tendency would be "static jumps - no iteration".
* increased throughput
* decreased latency per template regeneration (you know what I mean)